### ISSUE TITLE
Adding Pigeon Plink to patch crashes Rack.

### ISSUE THREAD

**Issue by ExistentiaVirae** [kockie69/RPJ#46]
Every time I attempt to add Pigeon Plink, Rack closes itself. These seem to be the pertinent lines in log.txt:

```
[182.703 info src/app/Browser.cpp:86 chooseModel] Creating module RPJ FM Pigeon Plink
[182.703 info src/PigeonPlink.cpp:373 onAdd] Loading config file /catacombs/Existentia/Rack/RPJ.json
[182.709 fatal adapters/standalone.cpp:49 fatalSignalHandler] Fatal signal 11. Stack trace:
20: /catacombs/Existentia/Rack2Free/Rack() [0x403c8d]
```

There is no RPJ.json file. I suspect that may be the root cause of the problem.

This was on Kubuntu 20.04 and VCV Rack Free v2.0.6

**Comment by kockie69** [comment]
The RPJ.json file is not needed, it is optional. You can add it yourself to add steppings ( or remove steppings). If it doesn't exist it will use the default settings that are hard-coded in the module.

 I checked for the handling of the message and that looks wrong in the code. Even though there is no RPJ.json file, it still says it is loading the file. So the cause of the crash is something else. 

Can you send me your complete logfile after a crash?

**Comment by spectromas** [comment]
Same thing happens for me, also on linux.
[log.txt](https://github.com/kockie69/RPJ/files/7918569/log.txt)

**Comment by kockie69** [comment]
@spectromas or @ExistentiaVirae , are you able (do you know how) to install the plugin manually so you can test a newer version?

If so, can you please pick aversion from here and install?
Can you send me the results please including the logfile if it still crashes?

**Comment by ExistentiaVirae** [comment]
[log.txt](https://github.com/kockie69/RPJ/files/7918613/log.txt)
I deleted the old log.txt and generated this one by adding Pigeon Plink to a new session.

I could build/install a copy manually to test, if you tell me which branch to clone.

**Comment by kockie69** [comment]
Sorry, forgot to add the link. You don't have to build, it is already done. You can pick it from here:

https://github.com/kockie69/RPJ/actions/runs/1733265194

Let me know if you need help

**Comment by spectromas** [comment]
Same thing for me with the beta as far as I can tell. 

```
[6.816 info src/app/Browser.cpp:87 chooseModel] Creating module RPJ FM Pigeon Plink
[6.816 info src/PigeonPlink.cpp:409 onAdd] Config file /home/alex/.Rack2/RPJ.json does not exist, using default settings
[6.817 fatal adapters/standalone.cpp:49 fatalSignalHandler] Fatal signal 11. Stack trace:
21: /home/alex/.local/share/VCV/Rack2Pro/Rack() [0x403ded]
20: /usr/lib/libc.so.6(+0x3cda0)
19: /usr/lib/libc.so.6(fclose+0xb)
18: /home/alex/.Rack2/plugins/RPJ/plugin.so(PigeonPlinkModule::onAdd(rack::engine::Module::AddEvent const&)+0x4c9)
17: ./libRack.so(rack::engine::Engine::addModule(rack::engine::Module*)+0x3b3)
16: ./libRack.so(+0x394466)
15: ./libRack.so(rack::app::browser::ModelBox::onButton(rack::widget::Widget::ButtonEvent const&)+0x90)
14: ./libRack.so(rack::widget::Widget::onButton(rack::widget::Widget::ButtonEvent const&)+0xed)
13: ./libRack.so(rack::widget::Widget::onButton(rack::widget::Widget::ButtonEvent const&)+0xed)
12: ./libRack.so(rack::widget::Widget::onButton(rack::widget::Widget::ButtonEvent const&)+0xed)
11: ./libRack.so(rack::ui::ScrollWidget::onButton(rack::widget::Widget::ButtonEvent const&)+0x103)
10: ./libRack.so(rack::app::browser::Browser::onButton(rack::widget::Widget::ButtonEvent const&)+0xfd)
9: ./libRack.so(rack::ui::MenuOverlay::onButton(rack::widget::Widget::ButtonEvent const&)+0x11d)
8: ./libRack.so(rack::widget::OpaqueWidget::onButton(rack::widget::Widget::ButtonEvent const&)+0x105)
7: ./libRack.so(rack::widget::EventState::handleButton(rack::math::Vec, int, int, int)+0x269)
6: ./libRack.so(+0x384447)
5: ./libRack.so(_glfwPlatformPollEvents+0x102c)
4: ./libRack.so(rack::window::Window::step()+0x65)
3: ./libRack.so(rack::window::Window::run()+0x28)
2: /home/alex/.local/share/VCV/Rack2Pro/Rack(main+0xc15)
1: /usr/lib/libc.so.6(__libc_start_main+0xd5)
0: /home/alex/.local/share/VCV/Rack2Pro/Rack() [0x403d19]
```

**Comment by ExistentiaVirae** [comment]
I got this:

```
[3.832 info src/window/Window.cpp:50 loadFile] Loaded font /catacombs/Existentia/Rack/plugins/voxglitch/res/ShareTechMono-Regular.ttf
[4.223 info src/network.cpp:147 requestJson] Requesting JSON GET https://api.vcvrack.com/version?edition=Free
[5.974 info src/network.cpp:147 requestJson] Requesting JSON GET https://api.vcvrack.com/user
[7.351 info src/app/Browser.cpp:86 chooseModel] Creating module RPJ FM Pigeon Plink
[7.351 info src/PigeonPlink.cpp:409 onAdd] Config file /catacombs/Existentia/Rack/RPJ.json does not exist, using default settings
[7.357 fatal adapters/standalone.cpp:49 fatalSignalHandler] Fatal signal 11. Stack trace:
20: /catacombs/Existentia/Rack2Free/Rack() [0x403c8d]
19: /lib/x86_64-linux-gnu/libc.so.6(+0x46210)
18: /lib/x86_64-linux-gnu/libc.so.6(fclose+0xb)
17: /catacombs/Existentia/Rack/plugins/RPJ/plugin.so(PigeonPlinkModule::onAdd(rack::engine::Module::AddEvent const&)+0x4c9)
16: ./libRack.so(rack::engine::Engine::addModule(rack::engine::Module*)+0x3ab)
15: ./libRack.so(+0x386e96)
14: ./libRack.so(rack::app::browser::ModelBox::onButton(rack::widget::Widget::ButtonEvent const&)+0x2a9)
13: ./libRack.so(rack::widget::Widget::onButton(rack::widget::Widget::ButtonEvent const&)+0xed)
12: ./libRack.so(rack::widget::Widget::onButton(rack::widget::Widget::ButtonEvent const&)+0xed)
11: ./libRack.so(rack::widget::Widget::onButton(rack::widget::Widget::ButtonEvent const&)+0xed)
10: ./libRack.so(rack::ui::ScrollWidget::onButton(rack::widget::Widget::ButtonEvent const&)+0x103)
9: ./libRack.so(rack::app::browser::Browser::onButton(rack::widget::Widget::ButtonEvent const&)+0xfd)
8: ./libRack.so(rack::ui::MenuOverlay::onButton(rack::widget::Widget::ButtonEvent const&)+0x11d)
7: ./libRack.so(rack::widget::OpaqueWidget::onButton(rack::widget::Widget::ButtonEvent const&)+0x105)
6: ./libRack.so(rack::widget::EventState::handleButton(rack::math::Vec, int, int, int)+0x269)
5: ./libRack.so(_glfwPlatformPollEvents+0xdfc)
4: ./libRack.so(rack::window::Window::step()+0x5c)
3: ./libRack.so(rack::window::Window::run()+0x28)
2: /catacombs/Existentia/Rack2Free/Rack(main+0xbe0)
1: /lib/x86_64-linux-gnu/libc.so.6(__libc_start_main+0xf3)
0: /catacombs/Existentia/Rack2Free/Rack() [0x403bb9]
```

**Comment by kockie69** [comment]
Ok, I have removed the possibility to add your own configs for now to see if that fixes the bug.
Both of your logs show onAdd as being the place where the issue starts which is the part that handles the config file.

Can you please try this version (sorry for the inconvenience):

https://github.com/kockie69/RPJ/actions/runs/1733330226

**Comment by spectromas** [comment]
No worries, happy to help! Still crashing: 
```

[4.549 info src/app/Browser.cpp:87 chooseModel] Creating module RPJ FM Pigeon Plink
[4.549 info src/app/Browser.cpp:91 chooseModel] Creating module widget RPJ FM Pigeon Plink
[4.551 fatal adapters/standalone.cpp:49 fatalSignalHandler] Fatal signal 6. Stack trace:
20: /home/alex/.local/share/VCV/Rack2Pro/Rack() [0x403ded]
19: /usr/lib/libc.so.6(+0x3cda0)
18: /usr/lib/libc.so.6(gsignal+0x142)
17: /usr/lib/libc.so.6(abort+0x116)
16: /usr/lib/libc.so.6(+0x26747)
15: /usr/lib/libc.so.6(+0x35616)
14: /home/alex/.Rack2/plugins/RPJ/plugin.so(std::_Function_handler<float (float), ObjectCache<float>::getExp2Ex()::{lambda(float)#1}>::_M_invoke(std::_Any_data const&, float&&)+0x249)
13: /home/alex/.Rack2/plugins/RPJ/plugin.so(std::_Function_handler<void (), WVCO<WidgetComposite>::init()::{lambda()#1}>::_M_invoke(std::_Any_data const&)+0xde)
12: /home/alex/.Rack2/plugins/RPJ/plugin.so(WVCO<WidgetComposite>::step()+0x154e)
11: /home/alex/.Rack2/plugins/RPJ/plugin.so(PigeonPlinkModule::step()+0x52)
10: ./libRack.so(rack::engine::Module::doProcess(rack::engine::Module::ProcessArgs const&)+0xa4c)
9: ./libRack.so(rack::engine::Engine::stepBlock(int)+0x779)
8: ./libRack.so(rack::audio::Device::processBuffer(float const*, int, float*, int, int)+0x180)
7: ./libRack.so(rack::RtAudioDevice::rtAudioCallback(void*, void*, unsigned int, double, unsigned int, void*)+0xa3)
6: ./libRack.so(RtApiJack::callbackEvent(unsigned long)+0x1e0)
5: ./libRack.so(+0x7b3431)
4: /usr/lib/libjack.so.0(+0x1367f)
3: /usr/lib/libjack.so.0(+0x12db8)
2: /usr/lib/libjack.so.0(+0x30f51)
1: /usr/lib/libpthread.so.0(+0x9259)
0: /usr/lib/libc.so.6(clone+0x43)
```

**Comment by kockie69** [comment]
Thanks for helping. Well , looks like we have a completely different error message now.

**Comment by ExistentiaVirae** [comment]
The latest build works fine for me.

**Comment by kockie69** [comment]
Can you both test this version ? 
@ExistentiaVirae , it might be crashing for you again, as I tried to bring in the config again...

https://github.com/kockie69/RPJ/actions/runs/1733411602

**Comment by kockie69** [comment]
@spectromas : Have you used Kitchen Sink before from Squinky Labs? Is that module working for you?

And can you let me know on what linux distro you are?

**Comment by ExistentiaVirae** [comment]
> Can you both test this version ? @ExistentiaVirae , it might be crashing for you again, as I tried to bring in the config again...
> 
> https://github.com/kockie69/RPJ/actions/runs/1733411602

Yeah, it died with this:

```
[10.919 info src/network.cpp:147 requestJson] Requesting JSON GET https://api.vcvrack.com/modules
[11.630 info src/app/Browser.cpp:86 chooseModel] Creating module RPJ FM Pigeon Plink
[11.631 info src/PigeonPlink.cpp:409 onAdd] Config file /catacombs/Existentia/Rack/RPJ.json does not exist, using default settings
[11.637 fatal adapters/standalone.cpp:49 fatalSignalHandler] Fatal signal 11. Stack trace:
20: /catacombs/Existentia/Rack2Free/Rack() [0x403c8d]
19: /lib/x86_64-linux-gnu/libc.so.6(+0x46210)
18: /lib/x86_64-linux-gnu/libc.so.6(fclose+0xb)
17: /catacombs/Existentia/Rack/plugins/RPJ/plugin.so(PigeonPlinkModule::onAdd(rack::engine::Module::AddEvent const&)+0x4e0)
16: ./libRack.so(rack::engine::Engine::addModule(rack::engine::Module*)+0x3ab)
15: ./libRack.so(+0x386e96)
14: ./libRack.so(rack::app::browser::ModelBox::onButton(rack::widget::Widget::ButtonEvent const&)+0x2a9)
13: ./libRack.so(rack::widget::Widget::onButton(rack::widget::Widget::ButtonEvent const&)+0xed)
12: ./libRack.so(rack::widget::Widget::onButton(rack::widget::Widget::ButtonEvent const&)+0xed)
11: ./libRack.so(rack::widget::Widget::onButton(rack::widget::Widget::ButtonEvent const&)+0xed)
10: ./libRack.so(rack::ui::ScrollWidget::onButton(rack::widget::Widget::ButtonEvent const&)+0x103)
9: ./libRack.so(rack::app::browser::Browser::onButton(rack::widget::Widget::ButtonEvent const&)+0xfd)
8: ./libRack.so(rack::ui::MenuOverlay::onButton(rack::widget::Widget::ButtonEvent const&)+0x11d)
7: ./libRack.so(rack::widget::OpaqueWidget::onButton(rack::widget::Widget::ButtonEvent const&)+0x105)
6: ./libRack.so(rack::widget::EventState::handleButton(rack::math::Vec, int, int, int)+0x269)
5: ./libRack.so(_glfwPlatformPollEvents+0xdfc)
4: ./libRack.so(rack::window::Window::step()+0x5c)
3: ./libRack.so(rack::window::Window::run()+0x28)
2: /catacombs/Existentia/Rack2Free/Rack(main+0xbe0)
1: /lib/x86_64-linux-gnu/libc.so.6(__libc_start_main+0xf3)
0: /catacombs/Existentia/Rack2Free/Rack() [0x403bb9]
```

**Comment by spectromas** [comment]
So the version without the config actually did work ok for me. I originally loaded it in the patch I had been using (an old v1 patch) and it crashed but after starting a new patch allowed me to load the module. 

Now with the last version I get this: 

```
[0.315 info src/engine/Engine.cpp:1395 fromJson] Creating module RPJ FM Pigeon Plink
[0.315 info src/engine/Module.cpp:173 fromJson] Patch created with RPJ v2.1.0-db45883, currently using v2.1.0-f2f78ab.
[0.315 info src/PigeonPlink.cpp:409 onAdd] Config file /home/alex/.Rack2/RPJ.json does not exist, using default settings
[0.315 fatal adapters/standalone.cpp:49 fatalSignalHandler] Fatal signal 11. Stack trace:
11: /home/alex/.local/share/VCV/Rack2Pro/Rack() [0x403ded]
10: /usr/lib/libc.so.6(+0x3cda0)
9: /usr/lib/libc.so.6(fclose+0xb)
8: /home/alex/.Rack2/plugins/RPJ/plugin.so(PigeonPlinkModule::onAdd(rack::engine::Module::AddEvent const&)+0x4e0)
7: ./libRack.so(rack::engine::Engine::addModule(rack::engine::Module*)+0x3b3)
6: ./libRack.so(rack::engine::Engine::fromJson(json_t*)+0x12f)
5: ./libRack.so(rack::patch::Manager::fromJson(json_t*)+0x1ee)
4: ./libRack.so(rack::patch::Manager::loadAutosave()+0xee)
3: ./libRack.so(rack::patch::Manager::launch(std::__cxx11::basic_string<char, std::char_traits<char>, std::allocator<char> >)+0x55)
2: /home/alex/.local/share/VCV/Rack2Pro/Rack(main+0x808)
1: /usr/lib/libc.so.6(__libc_start_main+0xd5)
0: /home/alex/.local/share/VCV/Rack2Pro/Rack() [0x403d19]
```

I'm using Arch and don't have any problems with Kitchen Sink.

**Comment by kockie69** [comment]
@spectromas , I asked that question because you had raised that error (but apparently it wasn't there)  and that message reflected to some Kitshen Sink code, that was the reason I was asking.
Okay, so I have rewritten the config code a bit, that version is ready to be tested soon hopefully.

**Comment by kockie69** [comment]
Alright, I have a new version that can be tested:

https://github.com/kockie69/RPJ/actions/runs/1733477380

Fingers crossed it works this time

**Comment by ExistentiaVirae** [comment]
> Alright, I have a new version that can be tested:
> 
> https://github.com/kockie69/RPJ/actions/runs/1733477380
> 
> Fingers crossed it works this time

It works as expected this time.

**Comment by kockie69** [comment]
Alright, lets wait for confirmation from @spectromas too. Hopefully it also works at that side....

**Comment by spectromas** [comment]
I'm not really sure what's happening. With this latest: previously had Rack crashed so I said yes to clearing the patch, then opened the module and no problem. Then I went to New to open my template and then Rack crashes when I try to load the module. 

So I tried again but this time starting with my template, the module loaded with no problem. However, I went to New to reopen my template and tried to load the module and it crashed. 

```
[12.179 info src/app/Browser.cpp:87 chooseModel] Creating module RPJ FM Pigeon Plink
[12.179 info src/PigeonPlink.cpp:411 onAdd] Config file /home/alex/.Rack2/RPJ.json does not exist, using default settings
[12.179 info src/app/Browser.cpp:91 chooseModel] Creating module widget RPJ FM Pigeon Plink
[12.182 fatal adapters/standalone.cpp:49 fatalSignalHandler] Fatal signal 6. Stack trace:
20: /home/alex/.local/share/VCV/Rack2Pro/Rack() [0x403ded]
19: /usr/lib/libc.so.6(+0x3cda0)
18: /usr/lib/libc.so.6(gsignal+0x142)
17: /usr/lib/libc.so.6(abort+0x116)
16: /usr/lib/libc.so.6(+0x26747)
15: /usr/lib/libc.so.6(+0x35616)
14: /home/alex/.Rack2/plugins/RPJ/plugin.so(std::_Function_handler<float (float), ObjectCache<float>::getExp2Ex()::{lambda(float)#1}>::_M_invoke(std::_Any_data const&, float&&)+0x249)
13: /home/alex/.Rack2/plugins/RPJ/plugin.so(std::_Function_handler<void (), WVCO<WidgetComposite>::init()::{lambda()#1}>::_M_invoke(std::_Any_data const&)+0xde)
12: /home/alex/.Rack2/plugins/RPJ/plugin.so(WVCO<WidgetComposite>::step()+0x154e)
11: /home/alex/.Rack2/plugins/RPJ/plugin.so(PigeonPlinkModule::step()+0x52)
10: ./libRack.so(rack::engine::Module::doProcess(rack::engine::Module::ProcessArgs const&)+0xa4c)
9: ./libRack.so(rack::engine::Engine::stepBlock(int)+0x779)
8: ./libRack.so(rack::audio::Device::processBuffer(float const*, int, float*, int, int)+0x180)
7: ./libRack.so(rack::RtAudioDevice::rtAudioCallback(void*, void*, unsigned int, double, unsigned int, void*)+0xa3)
6: ./libRack.so(RtApiJack::callbackEvent(unsigned long)+0x1e0)
5: ./libRack.so(+0x7b3431)
4: /usr/lib/libjack.so.0(+0x1367f)
3: /usr/lib/libjack.so.0(+0x12db8)
2: /usr/lib/libjack.so.0(+0x30f51)
1: /usr/lib/libpthread.so.0(+0x9259)
0: /usr/lib/libc.so.6(clone+0x43)
```

**Comment by kockie69** [comment]
I keep asking about Kitchen Sink, sorry :-) 
But is it in your template maybe? Or any other Squinky Labs modules in it? 
The error you get is pointing to getExp2Ex(), this is code from Squinky Labs. So, any Squinky in your template?

Is it possible to share your template?

**Comment by spectromas** [comment]
I don't have kitchen sink in my template but there are two Comp IIs. It does seem to be that that causes the crash.

**Comment by kockie69** [comment]
I think someone mentioned something about two compII's in a patch somewhere on the forum. It wasn't raised here as an issue so I did not yet dive into it (mainly because I was trying to get the Pigeon Plink released). 
So are you saying the PP issue is solved as long as you don't have 2 CompII's in your patch?

**Comment by spectromas** [comment]
> So are you saying the PP issue is solved as long as you don't have 2 CompII's in your patch?

Yes, it seems so. When the comps aren't in the patch I can't make it crash. Should I open a report on your squinky repo?

**Comment by kockie69** [comment]
Yes please. That also gives other the opportunity to follow the status. Thank you.

**Comment by kockie69** [comment]
This has been fixed and released in v2.1.1
